keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -0,0 +1,743 @@
+          .fetch()
+          .intoResultSet()
+      }
+
+      while (!found && rs.next()) {
+        try {
+          found = predicate.invoke(mapper.readValue(rs.getString("body")))
+        } catch (e: Exception) {
+          log.error("Failed reading message with fingerprint: ${rs.getString("fingerprint")} " +
+            "message: ${rs.getString("body")}", e)
+        }
+        lastId = rs.getString("id")
+      }
+    } while (!found && rs.row == limit)
+
+    return found
+  }
+
+  override fun poll(callback: (Message, () -> Unit) -> Unit) {
+    poll(1, callback)
+  }
+
+  /**
+   * TODO: Emit metrics: histogram of poll runtime, count of messages grabbed per poll, count of passes
+   */
+  override fun poll(maxMessages: Int, callback: (Message, () -> Unit) -> Unit) {
+    val now = clock.instant().toEpochMilli()
+    var changed = 0
+
+    /**
+     * Selects the primary key ulid's of up to ([maxMessages] * 3) ready and unlocked messages,
+     * sorted by delivery time.
+     *
+     * To minimize lock contention, this is a non-locking read. The id's returned may be
+     * locked or removed by another instance before we can acquire them. We read more id's
+     * than [maxMessages] and shuffle them to decrease the likelihood that multiple instances
+     * polling concurrently are all competing for the oldest ready messages when many more
+     * than [maxMessages] are read.
+     *
+     * Candidate rows are locked via an autocommit update query by primary key that will
+     * only modify unlocked rows. When (candidates > maxMessages), a sliding window is used
+     * to traverse the shuffled candidates, sized to (maxMessages - changed) with up-to 3
+     * attempts (and update queries) to grab [maxMessages].
+     *
+     * I.e. if maxMessage == 5 and
+     * candidates == [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].shuffle() == [9, 3, 7, 1, 10, 8, 5, 2, 6, 4]
+     *
+     * - pass1: attempts to claim [9, 3, 7, 1, 10], locks 3 messages
+     * - pass2: attempts to claim [8, 5], locks 1 message
+     * - pass3: attempt to claim [2], succeeds but if not, there are no further attempts
+     * - proceeds to process 5 messages locked via 3 update queries.
+     *
+     * This makes a trade-off between grabbing the maximum number of ready messages per poll cycle
+     * vs. minimizing [poll] runtime which is also critical to throughput. In testing a scenario
+     * with up-to 100k ready messages and 7 orca/keiko-sql instances with [fillExecutorEachCycle]
+     * enabled and 20 handler threads, instances were able to successfully grab [maxMessages] on
+     * each poll cycle > 94% of the time and no poll cycles came up empty handed.
+     *
+     * Note: this differs from other [Queue.poll] implementations in that the message
+     * body is only saved back to the db in order to increment [AttemptsAttribute] if
+     * [MaxAttemptsAttribute] has been set to a positive integer. Otherwise,
+     * [AttemptsAttribute] is unused.
+     */
```

Comment:
😍 😍 😍 😍 😍

Reply:
This comment is amazing. Thank you for such a detailed explanation!